### PR TITLE
JWT access/refresh tokens

### PR DESCRIPTION
**Pull request by Vladyslav317**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,17 +17,22 @@
   },
   "homepage": "https://github.com/ChechaValerii/intership-template#readme",
   "dependencies": {
-    "@hapi/joi": "^17.1.1",
+    "@hapi/joi": "17.1.1",
+    "bcrypt": "^5.0.1",
```
Comment (v-checha) [inline package.json]: please remove corners

File: src/components/Auth/index.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+
+const UserModel = require('../User/model');
+const AuthModel = require('./model');
+
+function generateAccessToken(user) {
```
Comment (v-checha) [inline src/components/Auth/index.js]: move business logic to service

File: src/components/Auth/index.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  return jwt.sign({ user }, process.env.ACCESS_TOKEN_SECRET, { expiresIn: '30s' });
+}
+
+function generateRefreshToken(user) {
```
Comment (v-checha) [inline src/components/Auth/index.js]: move business logic to service

File: src/components/Auth/index.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  return jwt.sign({ user }, process.env.REFRESH_TOKEN_SECRET);
+}
+
+async function verifyToken(req, res, next) {
```
Comment (v-checha) [inline src/components/Auth/index.js]: move to middleware or policy file

File: src/components/Auth/index.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  try {
+    const authHeader = await req.headers.authorization;
+    const token = await authHeader && authHeader.split(' ')[1];
+
+    if (token == null) {
+      return res.sendStatus(401);
```
Comment (v-checha) [inline src/components/Auth/index.js]: we must send only json format

File: src/components/Auth/index.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    jwt.verify(token, process.env.ACCESS_TOKEN_SECRET, (err, authData) => {
+      if (err) {
+        return res.sendStatus(403);
```
Comment (v-checha) [inline src/components/Auth/index.js]: the same

File: src/components/Auth/index.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+      }
+
+      req.authData = authData;
+
+      return null;
+    });
+  } catch (err) {
+    res.sendStatus(err);
```
Comment (v-checha) [inline src/components/Auth/index.js]: we must send only json format and if you try send error to client please check error.message or error.name

File: src/components/Auth/index.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  return next();
+}
+
+async function getUser(req, res, next) {
+  try {
+    const user = await UserModel.findOne({ fullName: 'John Hey' }).exec();
```
Comment (v-checha) [inline src/components/Auth/index.js]: query is hardcoded, please use req.query

File: src/components/Auth/index.js [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+    return res.json(user) || res.json();
```
Comment (v-checha) [inline src/components/Auth/index.js]: not clear why thats need

File: src/components/Auth/index.js [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  } catch (err) {
+    return next(err);
+  }
+}
+
+async function getTokens(req, res, next) {
+  try {
+    const user = await UserModel.findOne({ fullName: 'John Hey' }).exec();
```
Comment (v-checha) [inline src/components/Auth/index.js]: the same

File: src/components/Auth/index.js [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+    return res.json(user) || res.json();
+  } catch (err) {
+    return next(err);
+  }
+}
+
+async function getTokens(req, res, next) {
```
Comment (v-checha) [inline src/components/Auth/index.js]: rename to login

File: src/components/Auth/index.js [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    if (!user) {
+      return res.status(401).json({
+        message: 'Invalid credentials',
+      });
+    }
+
+    const accToken = generateAccessToken(user);
+    const refToken = generateRefreshToken(user);
+
+    const hashedPassword = await bcrypt.hash(user.password, 10);
```
Comment (v-checha) [inline src/components/Auth/index.js]: thats must be in sign-up method, not sign-in

File: src/components/Auth/index.js [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+      },
+    });
+  } catch (err) {
+    return next(err);
+  }
+}
+
+async function getRefreshToken(req, res, next) {
+  try {
+    const user = await AuthModel.findOne({ fullName: 'John Hey' }).exec();
```
Comment (v-checha) [inline src/components/Auth/index.js]: the same

File: src/components/Auth/index.js [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+    if (!user) {
+      res.status(401).json({
+        message: 'Invalid credentials',
```
Comment (v-checha) [inline src/components/Auth/index.js]: why thats invalid credentials if user not found?

File: src/components/Auth/model.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    email: {
+      type: String,
+      required: true,
+    },
+    password: {
+      type: String,
+      required: true,
+    },
+    accessToken: {
+      type: String,
```
Comment (v-checha) [inline src/components/Auth/model.js]: why do we need save access token?

File: src/components/Auth/model.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    },
+    refreshToken: {
+      type: String,
+    },
+  },
+  {
+    collection: 'authmodel',
```
Comment (v-checha) [inline src/components/Auth/model.js:33]: please put fields createdAt updatedAt

File: src/components/User/index.js [hunk 17]
```diff
@@ -34,33 +34,33 @@ async function findAll(req, res, next) {
...
+    });
+  } catch (error) {
+    if (error instanceof ValidationError) {
+      return res.status(422).json({
+        error: error.name,
+        details: error.message,
+      });
     }
+
+    res.status(500).json({
```
Comment (v-checha) [inline src/components/User/index.js]: you must send server error at error middleware

File: src/components/User/index.js [hunk 18]
```diff
@@ -71,33 +71,33 @@ async function findById(req, res, next) {
...
+    });
+  } catch (error) {
+    if (error instanceof ValidationError) {
+      return res.status(422).json({
+        message: error.name,
+        details: error.message,
+      });
     }
+
+    res.status(500).json({
```
Comment (v-checha) [inline src/components/User/index.js]: the same

File: src/components/User/index.js [hunk 19]
```diff
@@ -108,33 +108,33 @@ async function create(req, res, next) {
...
```
Comment (v-checha) [inline src/components/User/index.js]: the same

File: src/components/User/index.js [hunk 20]
```diff
@@ -108,33 +108,33 @@ async function create(req, res, next) {
...
-
-        return next(error);
+  try {
+    const { error } = UserValidation.updateById(req.body);
+
+    if (error) {
+      throw new ValidationError(error.details);
+    }
+
+    const updatedUser = await UserService.updateById(req.body.id, req.body);
```
Comment (v-checha) [inline src/components/User/index.js]: use after validation value 
` const { error, value } = UserValidation.updateById(req.body);`

File: src/components/User/index.js [hunk 21]
```diff
@@ -145,39 +145,39 @@ async function updateById(req, res, next) {
...
+    });
+  } catch (error) {
+    if (error instanceof ValidationError) {
+      return res.status(422).json({
+        message: error.name,
+        details: error.message,
+      });
     }
+
+    res.status(500).json({
```
Comment (v-checha) [inline src/components/User/index.js]: the same

File: src/components/User/index.js [hunk 22]
```diff
@@ -145,39 +145,39 @@ async function updateById(req, res, next) {
...
-
-        return next(error);
+  try {
+    const { error } = UserValidation.deleteById(req.body);
+
+    if (error) {
+      throw new ValidationError(error.details);
+    }
+
+    const deletedUser = await UserService.deleteById(req.body.id);
```
Comment (v-checha) [inline src/components/User/index.js]: the same

File: src/components/Auth/index.js [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+        accToken,
+        refToken,
+      },
+    });
+  } catch (err) {
+    return next(err);
+  }
+}
+
+async function getRefreshToken(req, res, next) {
```
Comment (v-checha) [inline src/components/Auth/index.js]: rename to refreshTokens

File: src/components/Auth/index.js [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+      res.sendStatus(401);
+    }
+
+    jwt.verify(refreshToken, process.env.REFRESH_TOKEN_SECRET, (err, authData) => {
+      if (err) {
+        return res.sendStatus(403);
+      }
+
+      const accToken = generateAccessToken({ fullName: authData.fullName });
+      return res.json({ accessToken: accToken });
```
Comment (v-checha) [inline src/components/Auth/index.js:72]: you must send two tokens each time access + refresh

Comment (Vladyslav317) [reply src/components/Auth/index.js:72]: I have a question. Why do I need to send 2 tokens if I only need a refresh token in order to see the info of my user again? I'm getting 2 tokens in login function...

**Pull request closed without merging**